policy: Policy Map Overflow Lockdown

nathanjsweet opened the pull request:
This logic has an option (`enable-policy-overflow-lockdown-endpoint`),
 "true" by default, so that it can be disabled.
    
When a policy map needs more entries than is
allocated for the bpf map, the endpoint is locked
down. The policy cache is reset and the policy map
is replaced with a set of default deny entries.
The endpoint is put into full default-deny for every protocol.
    
If the lockdown calls fail, then the endpoint will be
regenerated, and another lockdown will be attempted (until
the underlying map pressure is relieved).
    
The Map structure in the bpf package now has a
Pin method and the policymap now has a CreateUnpinned
method so that it can be initialized before it is
created.

Incremental deletion is now done first in cases where
incremental addition before deletion would overflow the
policy map.

```release-note
policy: Endpoints with policy map overflows are locked down, and all ingress and egress traffic is denied.
```

christarazi commented on pkg/endpoint/bpf.go:
```diff
@@ -1053,7 +1074,7 @@ type policyMapPressureUpdater interface {
 }
 
 func (e *Endpoint) updatePolicyMapPressureMetric() {
-	value := float64(e.realizedPolicy.GetPolicyMap().Len()) / float64(e.policyMap.MaxEntries())
+	value := float64(e.desiredPolicy.GetPolicyMap().Len()) / float64(e.policyMap.MaxEntries())
```
I think this is a separate change that we should separately review in another PR, right?

nathanjsweet replied:
The reason it is important here is that `realizedPolicy` is going to be brought down to ~6 entries, so the metric will be wrong during lockdown mode.

christarazi replied:
Does that mean we will now still report a policymap overflow via the metric even though technically the capacity of the policymap has completely changed? I feel like we might be better off with a new metric that provides visibility into the lockdown state of the map as well, rather than co-opting this BPF map pressure metric and repurposing it in this special case.

nathanjsweet replied:
The absolute capacity has changed in lockdown mode, but the effective capacity should be measured against what the capacity would be usually.

This change maintains how the metric works now. In main `realizedPolicy` will blindly accept new entries while bpf insertion fails, so `bpf_map_pressure` is expected to go above `1.0`.

If we don't change how this metric is calculated, the metric will falsely report a ratio of 6/16384 `bpf_map_pressure` in lockdown mode, which is technically correct but effectively incorrect.

Adding a new metric for lockdown might be a useful idea, but `bpf_map_pressure` already conveys the same information at values `>=1.0`.

tommyp1ckles replied:
fyi: realizedPolicy and desiredPolicy are the same map,  that's what this [commit](https://github.com/cilium/cilium/pull/28184/commits/e657f117c9d8090b275f80df530c0b6e210d22a6) was trying to address (but in the context of making the map pressure stay below 100%).

nathanjsweet replied:
> realizedPolicy and desiredPolicy are the same map

They can be, but during regeneration (and now lockdown) they are not.

ti-mo commented on pkg/maps/policymap/policymap.go:
```diff
@@ -441,6 +441,13 @@ func Create(path string) error {
 }
 
+// Create unpinned creates a policy map without pinning it to the filesystem
+func CreateUnpinned(path string) (*PolicyMap, error) {
```
Could the current uses of `policymap.Create()` be updated to be followed by a `Pin()` instead? There shouldn't be too many ways of creating a map.

nathanjsweet replied:
We could add an option to the current `Create` method, but that would involve updating many more callers. If the number of callers for `CreateUnpinned` grows in the codebase, we should consider consolidating the create logic in one place.

ti-mo replied:
Are there that many callers? 🤔 My point is CreateUnpinned() should be the default, with Pin() being an explicit action. Adding an option to Create() would be even worse. :)

joestringer replied:
I don't have a strong feeling either way for this; the kernel separates create vs. pin and traditionally inside Cilium we haven't. Traditionally Cilium always pinned, then to handle the replacement of another map the pattern was to:
1. Keep the existing map + pin in place
2. Create a new map + new pin (`foo_pending` map name + pin on filesystem)
3. Once the new map is populated, delete the old map pin path and move the new one into the old one's place
4. If something fails, move the new pin to `foo_failed`. Then we can introspect the content on failure.

That pattern was quite useful for introspection early in the project, but I'll admit that I don't recall using the `foo_pending` / `foo_failed` map paths for years.

But that said, what I do feel more strongly about is if we're going to change the Create internal API for unrelated codepaths to change the meaning to avoid pinning and force every user to deal with the details of pinning, please submit that as a separate PR. This PR is complex enough as it is without including unrelated datapath pinning changes.

ti-mo reviewed (approved):
I like the approach. :+1: Map.Pin() addition LGTM as well.

aanm commented on pkg/endpoint/bpf.go:
```diff
@@ -50,11 +54,28 @@ const (
 
 	syncPolicymapControllerGroup = controller.NewGroup("sync-policymap")
+
+	allKeys []policy.Key
```
Can you add a comment to this variable? It's not clear to me.

aanm commented on pkg/endpoint/bpf.go:
```diff
@@ -1228,13 +1269,44 @@ func (e *Endpoint) applyPolicyMapChanges() error {
+// endpointPolicyLockdown puts the endpoint policy map into a lockdown
+// mode. A new bpf policy map is populated with deny all traffic entries,
+// the old map is unpinned, and the new map is pinned.
+func (e *Endpoint) endpointPolicyLockdown() error {
```
This is a really tough situation to be. This means that if we add an allow all policy that selects all identities, this will break the connectivity for the pods selected by the policy.

Since we are creating bpf maps, any can't we create a map with a bigger size and then swap the maps?

joestringer replied:
I think that this can end up being a resource usage question for the user: Is it OK to consume more resources? If resizing the map fails, then we may still need some sort of lockdown option like this. Could the alternative with resizing the map potentially trigger an OOM and hence limit the agent's ability to mitigate issues in this scenario?

giorio94 commented on pkg/endpoint/bpf.go:
```diff
@@ -1228,13 +1269,44 @@ func (e *Endpoint) applyPolicyMapChanges() error {
+	}
+	e.policyMap = pM
+	e.realizedPolicy.SetPolicyMap(mapState)
+	if err = e.policyMap.Pin(e.policyMapPath()); err != nil {
```
There will be a time period in which no map is pinned to the given path. I guess that this should be fine, though, as AFAIU the datapath will just drop all traffic in this case. Is my understanding correct?

nathanjsweet replied:
That is correct, and the endpoint will regenerate, as policy sync will fail if this method fails.

giorio94 commented on pkg/policy/api/l4.go:
```diff
@@ -28,6 +28,12 @@ func (l4 L4Proto) IsAny() bool {
 
+// SupportedProtocols returns the currently supported protocols in the policy
+// engine, excluding "ANY".
+func SupportedProtocols() []L4Proto {
+	return []L4Proto{ProtoTCP, ProtoUDP, ProtoSCTP}
+}
```
Out of curiosity, why are ICMP{,v6} protocols not included here? That seems to contradict the comment. More in general, any reason not to use `ANY` as protocol in the deny rules in lockdown mode?

nathanjsweet replied:
We do not support selecting ICMP ([there is a way to do it](https://docs.cilium.io/en/latest/security/policy/language/#limit-icmp-icmpv6-types)). Only these protocols are supported for network policy selection when using the general PortoProtocol type.

giorio94 replied:
Sure. But I guess you may want to also block ICMP/ICMP6 traffic when entering in lockdown mode, otherwise traffic that was previously possibly blocked by the policy would then be allowed.

nathanjsweet replied:
No it won't. Explicit deny-entries, counter-intuitively, deny all other traffic as well implicitly. ICMP logic is separate from this.

giorio94 replied:
Ah gotcha. Thinking a bit more about this, I guess that even an empty map should do the trick (without the explicit denies), but those make it more explicit that traffic will be denied.

nathanjsweet replied:
> I guess that even an empty map should do the trick (without the explicit denies), but those make it more explicit that traffic will be denied.

We would need at least one deny-entry, but correct.

giorio94 replied:
I guess that's because the userspace logic inserts the allow all entry inside the map if no explicit allow/deny entries are present. Otherwise at the BPF level an empty map should drop everything IIUC.

nathanjsweet replied:
>  Otherwise at the BPF level an empty map should drop everything IIUC.

That depends on a number of config options. We want the explicitly entries.

giorio94 reviewed (approved):
AFAICT changes look reasonable. I've just left a couple of minor comment inline, mostly for my own curiosity.

As for the commit message, there seems to be an incomplete sentence:
> [...] deny entries and the policy cache is
endpoint is put into full default-deny for every protocol.

And finally clustermesh-related workflow failures are due to the kvstoremesh image removal, and should be fixed rebasing on top of main.

giorio94 commented on pkg/endpoint/bpf.go:
```diff
@@ -1228,13 +1269,44 @@ func (e *Endpoint) applyPolicyMapChanges() error {
+		return fmt.Errorf("failed to unpin previous map: %w", err)
+	}
+	e.policyMap = pM
+	e.realizedPolicy.SetPolicyMap(mapState)
+	if err = e.policyMap.Pin(e.policyMapPath()); err != nil {
```
Are these operations undone if at a certain point the endpoint exits from lockdown mode?

nathanjsweet replied:
These changes don't need to be undone because they are simple replacements of the map and map state. As soon as the endpoint exits lockdown mode it will automatically remove the deny-all entries because they are  not present in the `desiredPolicy` structure and add the required entries.

giorio94 replied:
Ah, right. So the new map just starts to be used as a replacement of the old one.

giorio94 commented on pkg/endpoint/bpf.go:
```diff
@@ -1183,6 +1205,25 @@ func (e *Endpoint) applyPolicyMapChanges() error {
 		}
 	}
 
+	if e.desiredPolicy.GetPolicyMap().Len() > policymap.MaxEntries {
```
Even if the total number is less than MaxEntries, the insertion below may still fail, because insertions are performed before removals. But I guess that is fine, because then removals are performed (as we don't return on first error), which means that eventually insertions should succeed.

nathanjsweet replied:
There is no insertion into the old maps below. New maps are created and replace the old maps.

giorio94 replied:
Ah, I was referring to the case in which `e.desiredPolicy.GetPolicyMap().Len()` is less than (but close to) MaxEntries, so the endpoint doesn't enter into lock-down mode. But a few insertions may still fail during the first attempt if they cause an overflow, compensated by the subsequent removal of other entries (if I'm not missing anything).

nathanjsweet replied:
Good catch. I added a section after lockdown mode that deletes changes first if the adds will overflow the map (temporarily).

giorio94 replied:
If I'm not mistaken, there could still be one issue caused by the explicit denies, when exiting from lockdown mode. Let's say `desiredPolicy.GetPolicyMap().Len() = MAX-3`, which theoretically fits into the map. But given that there are the additional 6 deny entries, the map will overflow, causing `applyPolicyMapChangesLocked` to return an error. And then the extra deny entries will never be removed, as the reconciliation is [performed afterwards](https://github.com/cilium/cilium/blob/pr%2Fnathanjsweet%2Fadd-lockdown-mode-for-policy-map-overflows/pkg/endpoint/bpf.go#L1353) only if the previous operation succeeded.

nathanjsweet replied:
> when exiting from lockdown mode. Let's say desiredPolicy.GetPolicyMap().Len() = MAX-3, which theoretically fits into the map. But given that there are the additional 6 deny entries, the map will overflow, causing applyPolicyMapChangesLocked to return an error.

The logic, as written, will delete entries first if the adds put `realizedPolicy` (which will account for the 3 entries) over the limit.

giorio94 replied:
Doesn't that account for desired policy changes only, while the reconciliation between desired and realized is performed in a subsequent stage? Otherwise I'm missing the purpose of [`syncPolicyMapsWith`](https://github.com/cilium/cilium/blob/pr%2Fnathanjsweet%2Fadd-lockdown-mode-for-policy-map-overflows/pkg/endpoint/bpf.go#L1353).

tommyp1ckles commented on daemon/cmd/daemon_main.go:
```diff
@@ -1135,6 +1135,9 @@ func InitGlobalFlags(cmd *cobra.Command, vp *viper.Viper) {
 	flags.MarkHidden(option.MaxInternalTimerDelay)
 	option.BindEnv(vp, option.MaxInternalTimerDelay)
 
+	flags.Bool(option.EnablePolicyOverflowEndpointLockdown, true, "When an endpoint's policy map overflows shutdown all (ingress and egress) network traffic for that endpoint.")
```
nit: `overflows shutdown` -> `overflows, shutdown`

tommyp1ckles replied:
Do we want to expose this via helm?

nathanjsweet replied:
> Do we want to expose this via helm?

No, people can use `extraConfig` or `extraArgs` if they want. Turning this off should be challenging.

squeed commented on pkg/endpoint/bpf.go:
```diff
@@ -1148,12 +1215,13 @@ func (e *Endpoint) ApplyPolicyMapChanges(proxyWaitGroup *completion.WaitGroup) e
+// applyPolicyMapChangesLocked applies any incremental policy map changes
+// collected on the desired policy. The method must be called with the
+// endpoint mutex held.
+func (e *Endpoint) applyPolicyMapChangesLocked() error {
```
Oof, reviewing this sent me down a huge rabbit hole. Unfortunately, this function is imprecise and subtly {broken? confusing?} - and it was before you got here - which is a big bummer. Unfortunately I think the lockdown code might fall afoul of this as well.

`applyPolicyMapChanges()` is, unfortunately, called in **three** contexts:

1. There is an incremental policy change. `e.desiredPolicy == e.realizedPolicy`, and everything is happy
2. The pod is regenerating, but the policy hasn't changed. `e.desiredPolicy == e.realizedPolicy`
3. The pod is regenerating, but the policy **has** changed. `e.desiredPolicy != e.realizedPolicy`. (This is the buggy state.)

The problem with `applyPolicyMapChanges()` is that it unconditionally writes the set of deltas *also* to `e.realizedPolicy()`. Which is weird; since the deltas are coming from `e.desiredPolicy`.

I'm pretty sure this is totally broken; these deltas are irrelevant to to the outgoing realized policy.

squeed replied:
Right, this is weird because in `e.addPolicyKey()` we're pushing the delta in to the bpf map as well as `e.realizedPolicy.mapState`.

squeed replied:
Okay, this is weird and probably broken but not relevant to the PR. Resolving.... for now...

squeed commented on pkg/endpoint/bpf.go:
```diff
@@ -1183,36 +1251,36 @@ func (e *Endpoint) applyPolicyMapChanges() error {
+	} else {
+		e.lockDown = false
+	}
+	// Add policy map entries before deleting to avoid transient drops. If there
```
Upon leaving lockdown, `e.desiredPolicy` may have lots of accumulated, already-applied changes to the MapState. We will need a full synchronization with the realized BPF state (and **not** `e.realizedPolicy`

squeed replied:
Expanded on this in a general comment.

nathanjsweet replied:
> lots of accumulated, already-applied

Lots of accumulated? Yes. Already applied? No.

aanm commented on pkg/policy/api/l4.go:
```diff
@@ -28,6 +28,12 @@ func (l4 L4Proto) IsAny() bool {
+// SupportedProtocols returns the currently supported protocols in the policy
+// engine, excluding "ANY".
+func SupportedProtocols() []L4Proto {
+	return []L4Proto{ProtoTCP, ProtoUDP, ProtoSCTP}
```
Why don't we have `ICMP` nor `ICMPv6`?

nathanjsweet replied:
We do not support selecting ICMP as protocol that can be allowed or denied. [ICMP gets its own unique field](https://docs.cilium.io/en/latest/security/policy/language/#limit-icmp-icmpv6-types) so that users can't accidentally block discovery messages. Only these protocols are supported for network policy selection when using the general PortoProtocol type.

aanm commented on pkg/endpoint/bpf.go:
```diff
@@ -1370,8 +1508,14 @@ func (e *Endpoint) syncPolicyMapWithDump() error {
+		// syncPolicyMapWithDump has succeeded, as the policy mapstate is where it should be.
+		// The dump will not be useful at this point, so returning here rather than returning
+		// after the dump is ideal.
+		if errors.Is(err, ErrPolicyEntryMaxExceeded) {
```
Why don't we log a debug message similar to line 1394?

aanm commented on pkg/endpoint/bpf.go:
```diff
@@ -1249,14 +1319,82 @@ func (e *Endpoint) applyPolicyMapChanges() error {
+		}
+	} else {
+		var outErr error
+		e.desiredPolicy.GetPolicyMap().ForEachDeny(func(k policy.Key, entry policy.MapStateEntry) bool {
```
If I follow this correctly we are inserting denies first, when are the allows? Maybe the misleading part is the `ForEachDeny` method name.

nathanjsweet replied:
Sorry. I made a mistake. There will never be allows.

aanm commented on pkg/endpoint/bpf.go:
```diff
@@ -1249,14 +1319,82 @@ func (e *Endpoint) applyPolicyMapChanges() error {
+			if entry.IsDeny {
+				err = pM.DenyKey(policymapKey)
+			} else {
+				err = pM.AllowKey(policymapKey, entry.AuthType.Uint8(), entry.ProxyPort)
```
If we are iterating `ForEachDeny`, does it have any `AllowKey`?

nathanjsweet replied:
No. I made a mistake.

christarazi commented on pkg/endpoint/bpf.go:
```diff
@@ -1249,14 +1327,99 @@ func (e *Endpoint) applyPolicyMapChanges() error {
+			return true
+		})
+		if outErr != nil {
+			return fmt.Errorf("failed to add all deny policies: %w", outErr)
```
```suggestion
			return fmt.Errorf("failed to add all deny mapstate entries: %w", outErr)
```

christarazi commented on pkg/endpoint/bpf.go:
```diff
@@ -1249,14 +1327,99 @@ func (e *Endpoint) applyPolicyMapChanges() error {
+		return fmt.Errorf("failed to pin new map to %q: %w", e.policyMapPath(), err)
+	}
+	e.updatePolicyMapPressureMetric()
+	e.setState(state, "Entries exceed BPF map limit.")
```
```suggestion
	e.setState(state, "Entries exceed BPF policymap limit.")
```

christarazi commented on pkg/endpoint/endpoint.go:
```diff
@@ -102,6 +102,18 @@ const (
+	// StateFullLockdown is used to set the endpoint is in a
+	// lockdown mode in which no ingress or egress traffic is allowed.
+	StateFullLockdown = State(models.EndpointStateFullLockdown)
+
```
I'd like to see unit tests for the state transitions between these new states and existing states.

christarazi commented on pkg/endpoint/bpf.go:
```diff
@@ -1204,36 +1270,48 @@ func (e *Endpoint) applyPolicyMapChanges() error {
-		if entry.IsRedirectEntry() {
-			entry.ProxyPort = e.realizedRedirects[policy.ProxyIDFromKey(e.ID, keyToAdd)]
+				logfields.EndpointID: e.ID,
+			}).WithError(err).Error("Failed to lockdown endpoint. The endpoint may be compromised, consider quarantining or shutting down this node.")
```
When this occurs, I'm assuming that the endpoint will be stuck in the new "locking-down" state? If an endpoint is stuck in that state, then it is considered "compromised" because Cilium cannot enforce lockdown mode on it? I think if that's the implication, then it needs to be clearer. The docs also need to be updated with this information, along with the new states.

nathanjsweet replied:
> I'm assuming that the endpoint will be stuck in the new "locking-down" state?

No. It is not stuck. Other reconciliation attempts can bring it out of this state, including the policy map being relieved of pressure.

> The docs also need to be updated with this information, along with the new states.

Fair enough.

nathanjsweet committed:
policy: Refactor ConsumeMapChanges to Return ChangeState

ConsumeMapChanges returns values based on ChangeState.
The value ChangeState can be used to revert changes,
using the existing RevertChanges method on MapState.

Signed-off-by: Nate Sweet <nathanjsweet@pm.me>

nathanjsweet committed:
policy: Add MapState DenyLen, Reset, etc

- Add a DenyLen method on MapState
- Add SupportedProtocols method API
- Add changes to ChangeState struct
- Add ResetMapChanges method to EndpointPolicy

Signed-off-by: Nate Sweet <nathanjsweet@pm.me>

nathanjsweet committed:
bpf: Add Pin Method

Add the Pin method, offered by the ebpf library,
so that Maps created Unpinned can be pinned.

Signed-off-by: Nate Sweet <nathanjsweet@pm.me>

nathanjsweet committed:
helm: Add Helm Option for Policy Map Overflow

"endpointLockdownOnMapOverflow" is the option
that maps to config option "enable-policy-overflow-endpoint-lockdown".

Signed-off-by: Nate Sweet <nathanjsweet@pm.me>

nathanjsweet committed:
option: Add Option to Manage Policy Map Overflow Behavior

Add an option (`enable-lockdown-endpoint-on-policy-overflow`),
set to "true" by default.

Signed-off-by: Nate Sweet <nathanjsweet@pm.me>

nathanjsweet committed:
policy: Policy Map Overflow Lockdown

This logic has an option (`enable-lockdown-endpoint-on-policy-overflow`),
set to "true" by default, so that it can be disabled.

If this new mode is disabled the endpoint will enter
a regeneration loop.

When a policy map needs more entries than is
allocated for the bpf map, the endpoint is locked
down. If the endpoint has deny policies that can still
fit into the policy map then they will be insert into
the policy map with all allows removed. If there are
no deny policies or the deny policies cannot fit into
the map then the policy cache is reset and the policy map
is replaced with a set of default deny entries.
The endpoint is put into full default-deny for every protocol
(except for ICMP).

If the lockdown calls fail, then the endpoint will be
regenerated, and another lockdown will be attempted (until
the underlying map pressure is relieved). If the lockdown
mode is still inserting deny policy entries (rather than
being in a "full" lockdown) further
adjustments to the map may still happen (in the form of
deny policies being added and/or removed).

Incremental deletion is now done first in cases where
incremental addition before deletion would overflow the
policy map.

Signed-off-by: Nate Sweet <nathanjsweet@pm.me>

nathanjsweet committed:
docs: Update Endpoint Documentation

Update Endpoint documentation to include lockdown
mode description and update the state diagram to
include lockdown states.

Signed-off-by: Nate Sweet <nathanjsweet@pm.me>

joestringer commented on Documentation/security/policy/lifecycle.rst:
```diff
@@ -110,3 +114,35 @@ those rules will be dropped.  Otherwise, if the policy enforcement
+policies of an endpoint exceeds the BPF policy map maximum it will put the
+endpoint into one of two lockdown modes ("full", and "soft").
+
+A "full" lockdown disables all egress and ingress traffic (that is, all
+network traffic). A "full" lockdown occurs when the number of policy
+entries required to accomodate an endpoint's applicable policies exceeds
+the BPF policy map maximum. Furthermore, a full lockdown will only happen
+when an endpoint has no applicable deny policies that can be enforced exclusively
+(over and against all of its allow policies) **or** the number of
+deny entries required to enforce a "soft" lockdown exceed the BPF policy map
+maximum.
+
+A "soft" lockdown exclusively enforces an endpoint's deny policies and drops
+all of the endpoint's allow policies.
```
Practically from a user perspective, how does full vs. soft lockdown differ?

If I consider "soft" lockdown, it exclusively enforces deny policies. If there are deny policies, then the whole endpoint is is default deny mode, so effectively that means all traffic will be denied. Is that right? That sounds like the same user-facing behaviour as full lockdown.

joestringer replied:
Maybe for context here too, the endpoint states are reflected in metrics as well. Even if internally we think it helps to break the states down because it simplifies the logic, I wonder whether that degree of complexity is meaningful and useful to users in the user-facing states as well. I guess it's a debugging utility so it's useful to directly match the underlying state. But in both scenarios, there's simply too many policy rules to effectively express and the only actions the user can take are either to change their policies or increase the map size limit?

joestringer commented on Documentation/security/policy/lifecycle.rst:
```diff
@@ -110,3 +114,35 @@ those rules will be dropped.  Otherwise, if the policy enforcement
+A "soft" lockdown exclusively enforces an endpoint's deny policies and drops
+all of the endpoint's allow policies.
+
+All other states can ephemerally override a lockdown if they are triggered
+by a lifecycle event (for example), but the endpoint will go back into lockdown
+immediately.
```
Do you have some examples of what you mean by lifecycle events? Presumably a Pod must first be deployed and then can go into lockdown (perhaps immediately as Cilium creates the Endpoint, if the policy cannot be expressed within the map limits). So then what other lifecycle events are there, I guess there's endpoint deletion but that shouldn't impact policy calculations :thinking: 

Let's say a Pod is in lockdown and a new policy is created which adds additional entries. Will the Pod remain in lockdown during that new policy lifecycle event, even though overall the endpoint should have no changes applied because the new policy cannot be expressed, similar to how it was before the new policy?

joestringer commented on pkg/endpoint/bpf.go:
```diff
@@ -49,11 +53,37 @@ const (
+	// endpoint policy is too full to accomodate a given set of changes.
+	ErrPolicyEntryMaxExceeded = errors.New("policy map max entries limit exceeded")
+
+	// ErrPolicyMaxBadState indicates that the BPF map backing the endpoint
+	// policy and that the endpoint is not in a ready state, so lockdown
```
`ErrPolicyMaxBadState indicates that the BPF map backing the endpoint policy` appears to be missing some description. Was a line in the middle here accidentally deleted or something?

joestringer commented on pkg/endpoint/endpoint.go:
```diff
@@ -1366,7 +1387,7 @@ func (e *Endpoint) setState(toState State, reason string) bool {
 	case StateReady:
 		switch toState {
-		case StateWaitingForIdentity, StateDisconnecting, StateWaitingToRegenerate, StateRestoring:
+		case StateWaitingForIdentity, StateDisconnecting, StateWaitingToRegenerate, StateRestoring, StateLockingDown:
```
(This line is fine)

Shouldn't this feature also allow regenerating -> {StateFullLockdown,StateSoftLockdown}?

joestringer commented on pkg/endpoint/endpoint.go:
```diff
@@ -102,6 +102,18 @@ const (
 	// StateRestoring is used to set the endpoint is being restored.
 	StateRestoring = State(models.EndpointStateRestoring)
 
+	// StateLockingDown is used to set the endpoint is attempting
+	// to lockdown.
+	StateLockingDown = State(models.EndpointStateLockingDown)
```
Could you describe a bit more the relationship between StateLockingDown and StateRegenerating?

Broadly I would expect that policy calculation occurs as part of StateRegenerating, so there would be no need to have a difference. However, AFAIK incremental policy updates are not currently worked into the Endpoint state machine so I suspect that incremental updates occur to Endpoints while the endpoint is in ready state. I'm wondering whether that's why StateLockingDown is separate, because the endpoint may transition into StateLockingDown as the result of an incremental policy update, but that lockdown may not be fully realized until the policymap is synced (?) which would occur either as part of regeneration or maybe as part of a separate controller.

joestringer commented on pkg/endpoint/bpf.go:
```diff
@@ -1169,24 +1252,84 @@ func (e *Endpoint) ApplyPolicyMapChanges(proxyWaitGroup *completion.WaitGroup) e
+			e.getLogger().WithFields(logrus.Fields{
+				logfields.EndpointID: e.ID,
+			}).WithError(err).Error("Failed to come out of lockdown successfully:" +
+				"The endpoint may be compromised, consider quarantining or shutting down this node.")
```
How do we know the endpoint may be compromised? I think that this message is trying to express that Cilium could not accurately calculate the policy implications for this specific endpoint, so we don't know if it's safe or not. However, I think this could give users the impression that Cilium has detected an active attacker in the environment. In such a case it's just as likely that there's a policy calculation bug.

Let's say there is such a bug and the user's entire cluster gets into this situation. The user will come to Cilium Slack and report that their cluster is compromised and Cilium is asking that they shut the entire cluster down. Is that an appropriate or feasible response?

Tracking down the errors that can occur in this case, it seems like the primary failure mode is that the map is full, which most likely would just be a bug that the capacity of the map mismatches the calculated usage?

I guess I'm wondering partly here whether the appropriate response should be to send the endpoint back into regenerating state to refresh all of the realized state, rather than pushing this to the user.

joestringer commented on pkg/endpoint/endpoint.go:
```diff
@@ -1410,6 +1431,12 @@ func (e *Endpoint) setState(toState State, reason string) bool {
 		}
+	case StateLockingDown, StateFullLockdown, StateSoftLockdown:
+		switch toState {
+		case StateWaitingForIdentity, StateDisconnecting, StateWaitingToRegenerate, StateRestoring,
+			StateLockingDown, StateFullLockdown, StateSoftLockdown:
```
Some of these transitions seem to be missing from the diagram, notably the ability to move to waiting-to-regenerate. I'm not _so_ worried about the others but I think it's important to recognize that a locked down endpoint may be locked down but also be pending a regeneration.

(This partly also makes me wonder whether the state machine is a useful representation of the lockdown property, because it seems like the endpoint could be locked down but also be in the queue to be regenerated or being actively regenerated, and the security properties of lockdown would hold even though the logic of what the regeneration does is slightly different)

joestringer replied:
To expand on the last part there: I think it's definitely useful to have a "lockdown" state separate from "ready" as a way to better signal to users that the endpoint is acting according to the design by Cilium. It's logical to me that when an endpoint's state is regenerated, that can either result in a completely green "ready" state where Cilium believes everything is operating correctly, or alternatively result in a brown/red "locked-down" state where Cilium thinks that some state cannot be reflected into the datapath and there may be traffic impacts.

joestringer commented on pkg/policy/l4.go:
```diff
@@ -479,6 +479,8 @@ type ChangeState struct {
 	Deletes Keys                  // deleted keys, if not nil
 	Old     map[Key]MapStateEntry // Old values of all modified or deleted keys, if not nil
+
+	changes []MapChange // changes slice used to create ChangeState
```
Why does this need to exist? Please document the motivation here and/or in the commit message.

joestringer commented on pkg/policy/mapstate.go:
```diff
@@ -1424,28 +1429,38 @@ func (mc *MapChanges) AccumulateMapChanges(cs CachedSelector, adds, deletes []id
+	return cs
+}
+
+func (mc *MapChanges) addChanges(changes []MapChange) {
+	mc.mutex.Lock()
+	defer mc.mutex.Unlock()
+	for _, ch := range changes {
+		mc.changes = append(mc.changes, ch)
+	}
 }
```
I'm finding it a bit hard to follow the `changes` abstraction here. The structure already has various fields to express the changes that are being made. With this commit there is a new `changes` subfield which references a subset of the changes(?) and the `addChanges()` function doesn't update the rest of the fields. Is the abstraction breaking down here, or are there more nuances to the types of changes and the intent of this new field that needs to be better documented?

joestringer commented on pkg/endpoint/bpf.go:
```diff
@@ -1249,14 +1371,111 @@ func (e *Endpoint) applyPolicyMapChanges() error {
+	e.policyMap = pM
+	e.realizedPolicy = policy.NewEndpointPolicy(e.policyGetter.GetPolicyRepository())
+	e.realizedPolicy.SetPolicyMap(mapState)
+	if err = e.policyMap.Pin(e.policyMapPath()); err != nil {
```
The datapath BPF program references a policymap by the fd of the policymap at the time of the BPF programs being loaded.

What this means is that if we create the new policymap and pin it to the filesystem, we have not actually updated the datapath yet, we've just prepared everything in place so that the datapath can now reference the newly created map from the filesystem. The datapath continues to use the old map until the BPF programs are reloaded.

It looks to me like there are two potential callers that come down into this path:
1. `regenerateBPF()` first reloads the BPF programs, then syncs the policymap which calls here. By creating a new policymap at this point in the process, it will not take effect until the next regeneration.
2. `endpointmanager.UpdatePolicyMaps()`. There is no BPF program reload from this path, so the changes would not take effect.

joestringer reviewed (changes requested):
I reviewed some of the broader strokes of this proposal, and I think that in general there's two sorts of concerns raised:

1. Do we need to express so many different states? Are we making the user-facing view into this feature more complicated than it needs to be?
2. The replacement of the policymaps doesn't seem to take effect as intended.

I didn't review the core logic of calculating the individual entries at this time. Given that there's new logic, I would expect to see some unit tests that validate the correctness, but at a glance I don't see those.
